### ISSUE TITLE
Hashing opcodes work on strings instead of byte arrays

### ISSUE THREAD

**Issue by vostrnad** [saving-satoshi/saving-satoshi-script#6]
In your implementation of the script interpreter, hashing opcodes pop a stack item, convert it to a string and then hash it. However, in Bitcoin the stack items are always byte arrays, and interpreting them as strings results in inconsistencies.

For example, the script `OP_0 OP_SHA256` should return the hash of an empty byte array, but it instead hashes the string "0".

**Comment by benalleng** [comment]
Good point! I've been thinking about how to migrate this to a more faithful representation of how bitcoin deals with bytes vs our admittedly largely abstracted string based logic
